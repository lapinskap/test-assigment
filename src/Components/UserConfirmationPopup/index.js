import PropTypes from 'prop-types';
import React from 'react';
import ReactDOM from 'react-dom';
import {
  Button, ModalHeader, ModalBody, ModalFooter,
} from 'reactstrap';

import Popup from '../Popup/popup';

export default function UserConfirmationPopup(
  {
    onCancel,
    onConfirm,
    isOpen,
    title,
    confirmLabel,
    cancelLabel,
    message,
  },
) {
  return (
    <Popup id="confirmation" isOpen={isOpen} toggle={onCancel}>
      <ModalHeader toggle={onCancel}>{title}</ModalHeader>
      <ModalBody>{message}</ModalBody>
      <ModalFooter>
        <Button data-t1="cancel" color="light" onClick={onCancel}>
          {cancelLabel}
        </Button>
        <Button data-t1="confirm" color="primary" onClick={onConfirm}>
          {confirmLabel}
        </Button>
      </ModalFooter>
    </Popup>
  );
}

/**
 * Used by Browser router
 * @param message
 * @param callback
 * @param title
 */
export function getUserConfirmationPopup(message, callback, title = 'Czy na pewno chcesz opuścić tę stronę?') {
  if (!message) {
    callback(true);
    return;
  }
  const container = document.createElement('div');
  container.setAttribute('custom-confirmation-navigation', '');
  document.body.appendChild(container);
  const closeModal = (callbackState) => {
    ReactDOM.unmountComponentAtNode(container);
    callback(callbackState);
  };
  ReactDOM.render(
    <UserConfirmationPopup
      title={title}
      onCancel={() => closeModal(false)}
      onConfirm={() => closeModal(true)}
      message={message}
    />,
    container,
  );
}

UserConfirmationPopup.propTypes = {
  cancelLabel: PropTypes.string,
  confirmLabel: PropTypes.string,
  isOpen: PropTypes.bool,
  message: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  onCancel: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
  title: PropTypes.string,
};

UserConfirmationPopup.defaultProps = {
  cancelLabel: 'Nie',
  confirmLabel: 'Tak',
  isOpen: true,
  message: 'Czy na pewno chcesz wykonać tę operację?',
  title: 'Czy jesteś pewien?',
};
