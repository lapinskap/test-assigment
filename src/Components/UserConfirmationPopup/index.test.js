import React from 'react';
import { mount, shallow } from 'enzyme';
import { act } from 'react-test-renderer';
import EnzymeToJson from 'enzyme-to-json';
import UserConfirmationPopup, { getUserConfirmationPopup } from './index';
import '../../tests/setupTests';

describe('getUserConfirmationPopup()', () => {
  it('executes getUserConfirmationPopup correctly on cancel click', () => {
    const callback = jest.fn((result) => {
      expect(result).toBe(false);
    });
    const message = 'something went wrong';
    getUserConfirmationPopup(message, callback);
    window.document.querySelectorAll('button')[0].click();
    expect(callback).toHaveBeenCalledTimes(1);
  });
  it('executes getUserConfirmationPopup correctly on confirm click', () => {
    const callback = jest.fn((result) => {
      expect(result).toBe(true);
    });
    const message = 'something went wrong';
    getUserConfirmationPopup(message, callback);

    window.document.querySelectorAll('button')[2].click();
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it('executes getUserConfirmationPopup without message', () => {
    const callback = (result) => {
      expect(result).toBe(true);
    };
    const message = '';
    getUserConfirmationPopup(message, callback);
  });
});

describe('<UserConfirmationPopup />', () => {
  const onCancel = () => ({});
  const onConfirm = () => ({});
  const isOpen = true;
  const title = 'Czy jesteś pewien?';
  const confirmLabel = 'Tak';
  const cancelLabel = 'Nie';
  const message = 'Czy na pewno chcesz wyknoać tę operację?';

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with default state', () => {
    const tree = mount(
      <UserConfirmationPopup
        onCancel={onCancel}
        onConfirm={onConfirm}
        isOpen={isOpen}
      />,
    );

    expect(EnzymeToJson(tree))
      .toMatchSnapshot();
  });

  it('should call onCancel correctly', async () => {
    const onCancelSpy = jest.fn();
    const wrapper = shallow(
      <UserConfirmationPopup
        onCancel={onCancelSpy}
        onConfirm={onConfirm}
        isOpen={isOpen}
        title={title}
        confirmLabel={confirmLabel}
        cancelLabel={cancelLabel}
        message={message}
      />,
    );
    const popupModal = wrapper.find('Button').at(0);
    await act(async () => {
      await popupModal.simulate('click');
    });
    wrapper.update();
    expect(onCancelSpy).toHaveBeenCalled();
  });

  it('should call onConfirm correctly', async () => {
    const onConfirmSpy = jest.fn();
    const wrapper = shallow(
      <UserConfirmationPopup
        onCancel={onCancel}
        onConfirm={onConfirmSpy}
        isOpen={isOpen}
        title={title}
        confirmLabel={confirmLabel}
        cancelLabel={cancelLabel}
        message={message}
      />,
    );
    const popupModal = wrapper.find('Button').at(1);
    await act(async () => {
      await popupModal.simulate('click');
    });
    wrapper.update();
    expect(onConfirmSpy).toHaveBeenCalled();
  });
});
