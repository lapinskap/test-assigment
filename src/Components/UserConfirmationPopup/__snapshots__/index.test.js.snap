// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserConfirmationPopup /> renders correctly with default state 1`] = `
<UserConfirmationPopup
  cancelLabel="Nie"
  confirmLabel="Tak"
  isOpen={true}
  message="Czy na pewno chcesz wykonać tę operację?"
  onCancel={[Function]}
  onConfirm={[Function]}
  title="Czy jesteś pewien?"
>
  <Popup
    id="confirmation"
    isOpen={true}
    toggle={[Function]}
    unmountOnClose={true}
  >
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      centered={false}
      container="body"
      fade={true}
      isOpen={true}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      returnFocusAfterClose={true}
      role="dialog"
      scrollable={false}
      toggle={[Function]}
      unmountOnClose={true}
      zIndex={1050}
    >
      <Portal
        node={
          <div
            style="position: relative; z-index: 1050; display: block;"
            tabindex="-1"
          >
            <div
              class=""
            >
              <div
                class="modal fade"
                role="dialog"
                style="display: block;"
                tabindex="-1"
              >
                <div
                  class="modal-dialog"
                  role="document"
                >
                  <div
                    class="modal-content"
                  >
                    <div
                      data-t1="confirmation"
                    >
                      <div
                        class="modal-header"
                      >
                        <h5
                          class="modal-title"
                        >
                          Czy jesteś pewien?
                        </h5>
                        <button
                          aria-label="Close"
                          class="close"
                          type="button"
                        >
                          <span
                            aria-hidden="true"
                          >
                            ×
                          </span>
                        </button>
                      </div>
                      <div
                        class="modal-body"
                      >
                        Czy na pewno chcesz wykonać tę operację?
                      </div>
                      <div
                        class="modal-footer"
                      >
                        <button
                          class="btn btn-light"
                          data-t1="cancel"
                          type="button"
                        >
                          Nie
                        </button>
                        <button
                          class="btn btn-primary"
                          data-t1="confirm"
                          type="button"
                        >
                          Tak
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="modal-backdrop fade"
              />
            </div>
          </div>
        }
      >
        <Portal
          containerInfo={
            <div
              style="position: relative; z-index: 1050; display: block;"
              tabindex="-1"
            >
              <div
                class=""
              >
                <div
                  class="modal fade"
                  role="dialog"
                  style="display: block;"
                  tabindex="-1"
                >
                  <div
                    class="modal-dialog"
                    role="document"
                  >
                    <div
                      class="modal-content"
                    >
                      <div
                        data-t1="confirmation"
                      >
                        <div
                          class="modal-header"
                        >
                          <h5
                            class="modal-title"
                          >
                            Czy jesteś pewien?
                          </h5>
                          <button
                            aria-label="Close"
                            class="close"
                            type="button"
                          >
                            <span
                              aria-hidden="true"
                            >
                              ×
                            </span>
                          </button>
                        </div>
                        <div
                          class="modal-body"
                        >
                          Czy na pewno chcesz wykonać tę operację?
                        </div>
                        <div
                          class="modal-footer"
                        >
                          <button
                            class="btn btn-light"
                            data-t1="cancel"
                            type="button"
                          >
                            Nie
                          </button>
                          <button
                            class="btn btn-primary"
                            data-t1="confirm"
                            type="button"
                          >
                            Tak
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="modal-backdrop fade"
                />
              </div>
            </div>
          }
        >
          <div
            className=""
          >
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onClick={[Function]}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              role="dialog"
              style={
                Object {
                  "display": "block",
                }
              }
              tabIndex="-1"
              tag="div"
              timeout={300}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={false}
              >
                <div
                  className="modal fade"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  role="dialog"
                  style={
                    Object {
                      "display": "block",
                    }
                  }
                  tabIndex="-1"
                >
                  <div
                    className="modal-dialog"
                    role="document"
                  >
                    <div
                      className="modal-content"
                    >
                      <div
                        data-t1="confirmation"
                      >
                        <ModalHeader
                          charCode={215}
                          closeAriaLabel="Close"
                          tag="h5"
                          toggle={[Function]}
                          wrapTag="div"
                        >
                          <div
                            className="modal-header"
                          >
                            <h5
                              className="modal-title"
                            >
                              Czy jesteś pewien?
                            </h5>
                            <button
                              aria-label="Close"
                              className="close"
                              onClick={[Function]}
                              type="button"
                            >
                              <span
                                aria-hidden="true"
                              >
                                ×
                              </span>
                            </button>
                          </div>
                        </ModalHeader>
                        <ModalBody
                          tag="div"
                        >
                          <div
                            className="modal-body"
                          >
                            Czy na pewno chcesz wykonać tę operację?
                          </div>
                        </ModalBody>
                        <ModalFooter
                          tag="div"
                        >
                          <div
                            className="modal-footer"
                          >
                            <Button
                              color="light"
                              data-t1="cancel"
                              onClick={[Function]}
                              tag="button"
                            >
                              <button
                                aria-label={null}
                                className="btn btn-light"
                                data-t1="cancel"
                                onClick={[Function]}
                                type="button"
                              >
                                Nie
                              </button>
                            </Button>
                            <Button
                              color="primary"
                              data-t1="confirm"
                              onClick={[Function]}
                              tag="button"
                            >
                              <button
                                aria-label={null}
                                className="btn btn-primary"
                                data-t1="confirm"
                                onClick={[Function]}
                                type="button"
                              >
                                Tak
                              </button>
                            </Button>
                          </div>
                        </ModalFooter>
                      </div>
                    </div>
                  </div>
                </div>
              </Transition>
            </Fade>
            <Fade
              appear={true}
              baseClass="fade"
              baseClassActive="show"
              className="modal-backdrop"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={true}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              tag="div"
              timeout={150}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={150}
                unmountOnExit={false}
              >
                <div
                  className="modal-backdrop fade"
                />
              </Transition>
            </Fade>
          </div>
        </Portal>
      </Portal>
    </Modal>
  </Popup>
</UserConfirmationPopup>
`;
